File: problems/096/096.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    int sudoku[50][9][9];

    int sum = 0;

    FILE *fp;
    char num_str[10];
    const char* FILE_NAME = "problems/096/p096_sudoku.txt";
    fp = fopen(FILE_NAME, "r");

    int index    = -1;
    int subindex =  0;

    while (fgets(num_str, 10, fp) != NULL)
    {
        if (isalpha(num_str[0]))
        {
            subindex = 0;
            index++;
            continue;
        }

        if (isdigit(num_str[0]))
        {
            int row_data[9];
            for (int i = 0; i < 9; i++)
                row_data[i] = num_str[i] - '0';
            memcpy(&sudoku[index][subindex++][0], &row_data[0], 9 * sizeof(int));
        }
    }

    for (int puzzle = 0; puzzle < 50; puzzle++)
    {
        int is_fixed_cell[9][9] = { 0 };

        for (int i = 0; i < 9; i++)
            for (int j = 0; j < 9; j++)
                if (sudoku[puzzle][i][j])
                    is_fixed_cell[i][j] = 1;

        int digit[9][9][10] = { 0 };

        int i = 0;
        while (i < 9)
        {
            int j = 0;
            while (j < 9)
            {
                if (is_fixed_cell[i][j])
                {
                    j++;
                    continue;
                }

                int i2 = (int)(i / 3) * 3;
                int j2 = (int)(j / 3) * 3;

                int duplicate_found = 1;

                if (!digit[i][j][1] || !digit[i][j][2] || !digit[i][j][3] 
                ||  !digit[i][j][4] || !digit[i][j][5] || !digit[i][j][6] 
                ||  !digit[i][j][7] || !digit[i][j][8] || !digit[i][j][9])
                {
                    for (int k = 1; k <= 9; k++)
                    {
                        if (digit[i][j][k])
                            continue;

                        duplicate_found = 0;

                        for (int l = 0; l < 9; l++)
                        {
                            if (l != i)
                            {
                                if (sudoku[puzzle][l][j] == k)
                                {
                                    duplicate_found = 1;
                                    break;
                                }
                            }

                            if (l != j)
                            {
                                if (sudoku[puzzle][i][l] == k)
                                {
                                    duplicate_found = 1;
                                    break;
                                }
                            }
                        }

                        if (!duplicate_found)
                        {
                            if ((sudoku[puzzle][i2][j2]     == k) || (sudoku[puzzle][i2][j2+1]   == k) 
                            ||  (sudoku[puzzle][i2][j2+2]   == k) || (sudoku[puzzle][i2+1][j2]   == k) 
                            ||  (sudoku[puzzle][i2+1][j2+1] == k) || (sudoku[puzzle][i2+1][j2+2] == k) 
                            ||  (sudoku[puzzle][i2+2][j2]   == k) || (sudoku[puzzle][i2+2][j2+1] == k) 
                            ||  (sudoku[puzzle][i2+2][j2+2] == k)) 
                                duplicate_found = 1;
                        }

                        if (!duplicate_found)
                        {
                            digit[i][j][k] = 1;
                            sudoku[puzzle][i][j] = k;
                            break;
                        }
                    }
                }

                if (duplicate_found)
                {
                    memset(&digit[i][j][0], 0, 10 * sizeof(int));

                    sudoku[puzzle][i][j] = 0;

                    j--;
                    if (j < 0)
                    {
                        j = 8;
                        i--;
                    }
                    while (is_fixed_cell[i][j])
                    {
                        j--;
                        if (j < 0)
                        {
                            j = 8;
                            i--;
                        }
                    }
                }
                else
                {
                    j++;
                }
            }

            i++;
        }

        sum += ( (sudoku[puzzle][0][0] * 100) 
            +    (sudoku[puzzle][0][1] *  10) 
            +     sudoku[puzzle][0][2]);
    }

    printf("%d\n", sum);

    return 0;
}

File: problems/096/096.f90
PROGRAM F096

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    INTEGER, DIMENSION(50, 9, 9) :: SUDOKU

    INTEGER :: SUM_VAL   = 0
    INTEGER :: INDEX_VAL = 0
    INTEGER :: SUB_INDEX = 1

    INTEGER :: I, I2, J, J2, K, L, PUZZLE

    LOGICAL :: DUPLICATE_FOUND

    LOGICAL, DIMENSION(9, 9, 9)   :: DIGIT
    LOGICAL, DIMENSION(9, 9)      :: IS_FIXED_CELL

    CHARACTER (LEN=30), PARAMETER :: FILENAME = "problems/096/p096_sudoku.txt"
    CHARACTER (LEN=30)            :: NUM_STR

    OPEN (2, FILE = FILENAME, STATUS = 'OLD')

    DO
        READ(2, *, END=10) NUM_STR

        IF (NUM_STR(1:1) < '0' .OR. NUM_STR(1:1) > '9') THEN
            SUB_INDEX = 1
            INDEX_VAL = INDEX_VAL + 1
            CYCLE

        END IF

        DO I = 1, 9
            SUDOKU(INDEX_VAL, SUB_INDEX, I) = IACHAR(NUM_STR(I:I)) - IACHAR('0')
        END DO
        SUB_INDEX = SUB_INDEX + 1
    END DO

    10 CLOSE(2)

    DO PUZZLE = 1, 50

        IS_FIXED_CELL = .FALSE.

        DO I = 1, 9
            DO J = 1, 9
                IF (SUDOKU(PUZZLE, I, J) /= 0) IS_FIXED_CELL(I, J) = .TRUE.
            END DO
        END DO

        DIGIT = .FALSE.

        I = 1
        DO WHILE (I <= 9)
            J = 1
            DO WHILE (J <= 9)

                IF (IS_FIXED_CELL(I, J)) THEN
                    J = J + 1
                    CYCLE
                END IF

                I2 = (INT((I - 1) / 3) * 3) + 1
                J2 = (INT((J - 1) / 3) * 3) + 1

                DUPLICATE_FOUND = .TRUE.

                IF (.NOT. DIGIT(I, J, 1)  .OR. .NOT. DIGIT(I, J, 2)            &
               .OR. .NOT. DIGIT(I, J, 3)  .OR. .NOT. DIGIT(I, J, 4)            &
               .OR. .NOT. DIGIT(I, J, 5)  .OR. .NOT. DIGIT(I, J, 6)            &
               .OR. .NOT. DIGIT(I, J, 7)  .OR. .NOT. DIGIT(I, J, 8)            &
               .OR. .NOT. DIGIT(I, J, 9)) THEN

                    DO K = 1, 9

                        IF (DIGIT(I, J, K)) CYCLE

                        DUPLICATE_FOUND = .FALSE.

                        DO L = 1, 9
                            IF (L /= I) THEN
                                IF (SUDOKU(PUZZLE, L, J) == K) THEN
                                    DUPLICATE_FOUND = .TRUE.
                                    EXIT
                                END IF
                            END IF
                            IF (L /= J) THEN
                                IF (SUDOKU(PUZZLE, I, L) == K) THEN
                                    DUPLICATE_FOUND = .TRUE.
                                    EXIT
                                END IF
                            END IF
                        END DO

                        IF (.NOT. DUPLICATE_FOUND) THEN
                            IF ((SUDOKU(PUZZLE, I2,     J2)     == K)          &
                           .OR. (SUDOKU(PUZZLE, I2,     J2 + 1) == K)          &
                           .OR. (SUDOKU(PUZZLE, I2,     J2 + 2) == K)          &
                           .OR. (SUDOKU(PUZZLE, I2 + 1, J2)     == K)          &
                           .OR. (SUDOKU(PUZZLE, I2 + 1, J2 + 1) == K)          &
                           .OR. (SUDOKU(PUZZLE, I2 + 1, J2 + 2) == K)          &
                           .OR. (SUDOKU(PUZZLE, I2 + 2, J2)     == K)          &
                           .OR. (SUDOKU(PUZZLE, I2 + 2, J2 + 1) == K)          &
                           .OR. (SUDOKU(PUZZLE, I2 + 2, J2 + 2) == K)) THEN

                                DUPLICATE_FOUND = .TRUE.

                            END IF
                        END IF

                        IF (.NOT. DUPLICATE_FOUND) THEN
                            DIGIT(I, J, K)  = .TRUE.
                            SUDOKU(PUZZLE, I, J) = K
                            EXIT
                        END IF

                    END DO

                END IF

                IF (DUPLICATE_FOUND) THEN

                    DIGIT(I, J, 1:9) = .FALSE.
                    SUDOKU(PUZZLE, I, J) = 0

                    J = J - 1
                    IF (J == 0) THEN
                        J = 9
                        I = I - 1
                    END IF

                    DO WHILE (IS_FIXED_CELL(I, J))
                        J = J - 1
                        IF (J == 0) THEN
                            J = 9
                            I = I - 1
                        END IF
                    END DO

                ELSE
                    J = J + 1
                END IF

            END DO
            I = I + 1
        END DO
        SUM_VAL = SUM_VAL + (SUDOKU(PUZZLE, 1, 1) * 100) +                     &
                            (SUDOKU(PUZZLE, 1, 2) * 10)  +                     &
                            (SUDOKU(PUZZLE, 1, 3))
    END DO

    PRINT ('(I0)'), SUM_VAL

END PROGRAM F096
